### ISSUE TITLE
The Twisted Coding Standard does not reflect current best practice

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#8256]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)| @hawkowl reported|
|-|-|
|Trac ID|trac#8256|
|Type|enhancement|
|Created|2016-03-26 12:16:44Z|
|Branch|https://github.com/twisted/twisted/tree/codingstandard-twentysixteen-8256|

Things it doesn't mention:

- non-global reactor imports
- the long-import lines says to put them on multiple lines, Glyph says ```from foo import (x, y,\n z)``` is acceptable
- *EVERYTHING* must be done on a branch
- It mentions Lore
- It says something about "adding to the setup.py", this isn't a thing anymore
- The Python 3 porting guidelines should be merged in to this.

<details><summary>Searchable metadata</summary>

```
trac-id__8256 8256
type__enhancement enhancement
reporter__hawkowl hawkowl
priority__normal normal
milestone__None None
branch__branches_codingstandard_twentysixteen_8256 branches/codingstandard-twentysixteen-8256
branch_author__hawkowl hawkowl
status__closed closed
resolution__fixed fixed
component__core core
keywords__None None
time__1458994604549711 1458994604549711
changetime__1461298550425477 1461298550425477
version__None None
owner__hawkowl hawkowl

```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_1"></a>|@hawkowl commented|
|-|-|

(In [47085]) Branching to codingstandard-twentysixteen-8256.

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_2"></a>|@hawkowl commented|
|-|-|

I think this is in a good spot right now.

I've tried to amend it to match the current "best practice" from code reviews, etc.

Please review.

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_3"></a>|@hawkowl commented|
|-|-|

For reviewers: This diff contains the major material changes:

https://github.com/twisted/twisted/compare/b4c6ccea0e82c31c3ae5a7eb55a09727100ecde1...codingstandard-twentysixteen-8256

**Comment by twisted-trac** [comment]
|[<img alt="radix's avatar" src="https://avatars.githubusercontent.com/u/227068?s=50" width="50" height="50">](https://github.com/radix)<a name="note_4"></a>|@radix commented|
|-|-|

> All strings in Twisted which are not interfacing directly with Python should be marked explicitly as "bytestrings" or "text/Unicode strings".

1. What does "interfacing directly with Python" mean here?


> many tests wish to provide their own reactor which controls the passage of time and simulate timeouts.

2. minor grammatical inconsistency here: simulate should be "simulates" (like "controls")

**Comment by twisted-trac** [comment]
|[<img alt="radix's avatar" src="https://avatars.githubusercontent.com/u/227068?s=50" width="50" height="50">](https://github.com/radix)<a name="note_5"></a>|@radix commented|
|-|-|

you should probably resubmit after addressing those points since I'm pretty out of touch with Twisted dev :(

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_6"></a>|@hawkowl commented|
|-|-|

Thanks radix, I've fixed those items :)

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_7"></a>|@adiroiban set owner to @hawkowl|
|-|-|


Many thanks for the update.

Here are my minor comments.

--------------

`Try to overloaded terms.` ... I think that we need an example here as the following text does not help me :)

----

not sure what to say about *alwasy* TDD. When I developer a prototype or try to validate an idea, I just draft some minimal code to play with.

For bug-fixing I agree that we should always followd TDD.

Also, for infra related code and devtools code, since those tools are not used by end users and are alwasy used by developers and the continous test system, I don't see any value in writing extra tests.

-----

Do we need to talk about `test_*` naming convention here? 

Same for how to run individal tests or associted test.

I would say that this info should only be found in the dedicated documentation.

This documentation page is already quite long.

----

Does a developer still need to learn Emacs ?

I feel that this guide is too opinionated here :) I like notepad.exe :p


---------

For function responding to events I also use _lcMethodName for methods called from  a loop cal ... and _tcMethodName for the one called from a thread. Not sure if it make sense to add this info.

-------

Regarding `Using the Global Reactor` since in the example the reactor is used as IReactorTime, maybe the private name should be 'clock'/'scheduler' instead of 'reactor'.


I agree that the global reactor should not be used for getting the time :)

I think that we should find a better example for avoiding using the global reactor.... maybe some code which works in parallel with 2 different reactor.


-------

Please check my comments and merge.

Thanks!

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_8"></a>|@hawkowl set status to `closed`|
|-|-|

(In [47302]) Merge codingstandard-twentysixteen-8256: Update the Twisted coding standard to reflect current best practice.

Author: hawkowl
Reviewers: radix, adiroiban
Fixes: [#8256](https://github.com/twisted/twisted/issues/8256)
